### PR TITLE
Transformers v4.12.0 compatible

### PR DESCRIPTION
**Pull request by JulianneKnott**
Updating Fastseq for compatibility with huggingface transformers v4.12.0. 
Benchmarks (samples/sec):
```
Bart       13.0 (2.9x)
DistilBart 19.1 (3.5x)
T5         29.2 (2.0x)
GPT2       20.3 (5.1x)
ProphetNet TBD
```

### CODE REVIEW

File: fastseq/optimizer/transformers/modeling_bart_optimizer.py [hunk 1]
```diff
@@ -3,247 +3,170 @@
...
-        attn_output = attn_output.transpose(0, 1).contiguous().view(
-            tgt_len, bsz, embed_dim)
-        attn_output = self.out_proj(attn_output)
         if output_attentions:
-            attn_weights = attn_weights.view(bsz, self.num_heads, tgt_len,
-                                             src_len)
+            # this operation is a bit awkward, but it's required to
+            # make sure that attn_weights keeps its gradient.
+            # In order to do so, attn_weights have to be reshaped
+            # twice and have to be reused in the following
```
Comment (feihugis) [inline fastseq/optimizer/transformers/modeling_bart_optimizer.py:147]: Could you explain why attn_weights need to be reshaped twice for keeping its gradient?

File: fastseq/optimizer/transformers/modeling_bart_optimizer.py [hunk 2]
```diff
@@ -3,247 +3,170 @@
...
-
-                reordered_past.append(layer_past_new)
-
-        new_enc_mask = (enc_mask if (enc_mask is None or enc_mask == [])
-            else enc_mask.index_select(0, beam_idx))
-
-        past = ((enc_out, new_enc_mask, *encoder_outputs[2:]), reordered_past)
-        return past
-
-MODEL_FOR_SEQ_TO_SEQ_CAUSAL_LM_MAPPING[BartConfig] = BartForConditionalGenerationV2 # pylint: disable=line-too-long
```
Comment (feihugis) [inline fastseq/optimizer/transformers/modeling_bart_optimizer.py:249]: Why is this removed? I think this is used to make `@replace` works as expected when loading Bart model. Maybe the latest hf will not need it, but had better to check if all the replaced functions have been used during the inference.

File: tests/optimizer/transformers/data/expected_prophetnet_output_no_cache.hypo [hunk 3]
```diff
@@ -0,0 +1,16 @@
+[SEP]'' it's a very disturbing scene.'' he added, `'so far,'' the prosecutor said, adding, ` ` so far... so far'' said the prosecutor, who told cnn on wednesday that he is not aware of any such video.'he said. the prosecutor also said that he has not seen any videos, he said, but that the person who has such a video is not being aware of the video. the publication said that the video was recovered at the site, but he said that was not the case. the reports by the two publications said that they are not. the publications said they did not have the video, and the two magazines said that neither of the publications had the video and that the publications did not know that the videos were found. the magazine said that no such video was found. asked that the report is not a video, the publications told cnn. [SEP] [PAD] [PAD] [PAD] [PAD]
```
Comment (feihugis) [inline tests/optimizer/transformers/data/expected_prophetnet_output_no_cache.hypo:1]: What are the differences between this file and the above one?

**Review by feihugis — COMMENTED** [review]
@JulianneKnott Thanks for updating HF models! I did a pass and left some minor comments. Did you check if all the changed code have been used during the inference?

File: fastseq/optimizer/transformers/modeling_bart_optimizer.py (hunk 1 above)
Comment (JulianneKnott) [reply fastseq/optimizer/transformers/modeling_bart_optimizer.py:147]: I am not sure. The comment comes from a change that was made in the latest huggingface transformers: https://github.com/huggingface/transformers/blob/69e16abf98c94b8a6d2cf7d60ca36f13e4fbee58/src/transformers/models/bart/modeling_bart.py#L240

File: tests/optimizer/transformers/data/expected_prophetnet_output_no_cache.hypo (hunk 3 above)
Comment (JulianneKnott) [reply tests/optimizer/transformers/data/expected_prophetnet_output_no_cache.hypo:1]: This is vestigial from some debug/sanity checking I was doing. I will remove it. Essentially, prophetnet has slightly different outputs depending on if it is with or without cache which I was checking, but since it takes 30 mins to run the test without cache on all 128 samples, this file is shorter so I could run it faster.

File: examples/bart/README.md [hunk 4]
```diff
@@ -61,7 +61,7 @@ Refer to [file](../../tests/optimizer/fairseq/test_fairseq_optimizer.py).
 
   |      BatchSize      |       32      |       64       |       128      |
   |:-------------------:|:-------------:|:--------------:|:--------------:|
-  | transformers-3.0.2  | 2.5 samples/s |      OOM       |      OOM       |
+  | transformers-4.12.0  | 2.5 samples/s |      OOM       |      OOM       |
```
[the remark this answers is not in the document]
Comment (JulianneKnott) [reply examples/bart/README.md:64]: Forgot to update the benchmarks in the examples. Will add them today.

File: fastseq/optimizer/transformers/beam_search_optimizer.py [hunk 5]
```diff
@@ -115,915 +179,1018 @@ def _update_beam_size(self, num_beams):
...
+                layer.cross_attn.num_beams = num_beams
+            logger.debug("num_beams has been updated to {}".format(num_beams))
+            return
+
         logger.debug(
             "The num_beams optimization in self_attn and encoder_decoder_attn "
             "does not support {} yet.".format(self.__class__))
 
 
+    def _get_logits_processor(
```
[the remark this answers is not in the document]
Comment (JulianneKnott) [reply fastseq/optimizer/transformers/beam_search_optimizer.py:194]: It replaces: https://github.com/huggingface/transformers/blob/69e16abf98c94b8a6d2cf7d60ca36f13e4fbee58/src/transformers/generation_utils.py#L559. It should be getting replaced when GenerationMixinV2 replaces GenerationMixin

File: fastseq/optimizer/transformers/beam_search_optimizer.py [hunk 6]
```diff
@@ -115,915 +179,1018 @@ def _update_beam_size(self, num_beams):
             logger.debug("num_beams has been updated to {}".format(num_beams))
             return
 
+        if isinstance(self, ProphetNetForConditionalGeneration):
+            for layer in self.prophetnet.decoder.layers:
+                #layer.self_attn.num_beams = num_beams
+                layer.cross_attn.num_beams = num_beams
```
[the remark this answers is not in the document]
Comment (JulianneKnott) [reply fastseq/optimizer/transformers/beam_search_optimizer.py:185]: No, forgot to remove it. Only cross_attn is optimized in ProphetNet

File: fastseq/optimizer/transformers/beam_search_optimizer.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (feihugis) [reply fastseq/optimizer/transformers/beam_search_optimizer.py:193]: Is `@replace(...)` needed here?

File: tests/optimizer/transformers/data/expected_prophetnet_output_no_cache.hypo (hunk 3 above)
Comment (feihugis) [reply tests/optimizer/transformers/data/expected_prophetnet_output_no_cache.hypo:1]: make sense

File: fastseq/optimizer/transformers/beam_search_optimizer.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (JulianneKnott) [reply fastseq/optimizer/transformers/beam_search_optimizer.py:193]: I don't think so, since it is a member of GenerationMixin which is already replaced here: https://github.com/microsoft/fastseq/blob/c3c98cbc10e13f7bcb7c456b3531d7438fafa948/fastseq/optimizer/transformers/beam_search_optimizer.py#L133

File: benchmarks/models/hf_t5.sh [hunk 7]
```diff
@@ -15,27 +15,24 @@ source hf.sh
     val \
     64 \
     --task translation_en_to_ro 
-#    --no_repeat_ngram_size 3	# baseline don't support this arg now.
```
[the remark this answers is not in the document]
Comment (JulianneKnott) [reply benchmarks/models/hf_t5.sh:18]: Looks like this parameter is still valid. Not sure why it was commented out originally. Adding it back.

File: benchmarks/hf.sh [hunk 8]
```diff
@@ -1,10 +1,9 @@
 #!/bin/bash
 source utils.sh
 if [[ $SKIP_BASELINE -eq 0 ]]; then
-    export BASELINE_REPO=$CACHE_DIR/transformers_v3.0.2
-    #https://github.com/huggingface/transformers.git \
+    export BASELINE_REPO=$CACHE_DIR/transformers_v4.12.0
     git_clone_if_not_in_cache \
-	https://github.com/JiushengChen/transformers.git \
```
Comment (JiushengChen) [inline benchmarks/hf.sh:7]: Some context about this forked repo, it is to add param "no_repeat_ngram_size", see
https://github.com/JiushengChen/Transformers/commit/db970431a6d55d5b1477e21ff719612e425c5f95

File: azure-pipelines.yml [hunk 9]
```diff
@@ -54,10 +60,10 @@ jobs:
         python -c "import torch; print('CUDA available:', torch.cuda.is_available())"
 
         echo "******* Running fairseq unittests *******"
-        bash tests/run_fairseq_tests.sh
+        # bash tests/run_fairseq_tests.sh
```
[the remark this answers is not in the document]
Comment (JulianneKnott) [reply azure-pipelines.yml:63]: This is temporary to run only the transformers tests in the pipeline. I will add it back once I have fixed the transformers tests so they run in the pipeline correctly.

File: benchmarks/hf.sh (hunk 8 above)
Comment (JulianneKnott) [reply benchmarks/hf.sh:7]: To clarify, this is no longer needed since fastseq uses it's own run_eval_hf.py for the baseline?

File: examples/bart/README.md [hunk 10]
```diff
@@ -61,8 +61,8 @@ Refer to [file](../../tests/optimizer/fairseq/test_fairseq_optimizer.py).
 
   |      BatchSize      |       32      |       64       |       128      |
   |:-------------------:|:-------------:|:--------------:|:--------------:|
-  | transformers-3.0.2  | 2.5 samples/s |      OOM       |      OOM       |
-  |  above + fastseq    | 7.6 samples/s | 11.3 samples/s  | 12.4 samples/s  |
+  | transformers-4.12.0  | 4.5 samples/s | 4.5 samples/s |      OOM       |
+  |  above + fastseq    | 10.6 samples/s | 10.6 samples/s  | 12.4 samples/s  |
```
Comment (feihugis) [inline examples/bart/README.md:65]: Why do BatchSize = 32 and = 64 for fastseq have the same speed?

File: examples/gpt2/README.md [hunk 11]
```diff
@@ -6,10 +6,10 @@ OpenAI GPT-2 model was proposed in [Language Models are Unsupervised Multitask L
 
 - Speed on single NVIDIA-V100-16GB
 
-  |       BatchSize       |        64       |      128       |
-  |:---------------------:|:---------------:|:--------------:|
-  |   transformers_v3.0.2 |   3.0 samples/s |      OOM       |
-  |   above + fastseq     |  11.2 samples/s | 16.5 samples/s |
+  |       BatchSize       |        64       |      128       |      256       |
+  |:---------------------:|:---------------:|:--------------:|:--------------:|
+  |   transformers_v4.12.0 |   3.9 samples/s |      3.8 samples/s       |      OOM       |
```
Comment (feihugis) [inline examples/gpt2/README.md:11]: for transformer, BatchSize = 128 is slower than 64. Did we double check it?

File: examples/prophetnet/README.md [hunk 12]
```diff
@@ -48,3 +48,40 @@ Refer to [file](../../tests/models/test_prophetnet_fs.py).
 ```bash
 bash generate_binary_data_for_prophetnet.sh INPUT_DATA_DIR
 ```
+
+## Speedup ProphetNet (Huggingface Transformers version) by using FastSeq
+
+- CNN daily mail validation data, NVIDIA-V100-16GB
+
+  |      BatchSize      |       128      |
```
Comment (feihugis) [inline examples/prophetnet/README.md:56]: Could we add more batchsize here as the  above fairseq experiments?

File: fastseq/config.py [hunk 13]
```diff
@@ -18,8 +18,8 @@
 FASTSEQ_VERSION = '0.2.0'
 
 # supported versions of transformers
-MIN_TRANSFORMERS_VERSION = '3.0.2'
-MAX_TRANSFORMER_VERSION = '3.0.2'
+MIN_TRANSFORMERS_VERSION = '4.12.0'
+MAX_TRANSFORMER_VERSION = '4.12.0'
```
Comment (feihugis) [inline fastseq/config.py:22]: ditto

File: tests/run_fastseq_tests.sh [hunk 14]
```diff
@@ -1,11 +1,13 @@
...
-
+rm -rf build/
+rm ngram_repeat_block_cuda*.so
+pip install --editable .
 for fastseq_py_test_file in $(find tests/ -name "test_*.py")
 do 
   if [[ $fastseq_py_test_file == *"test_fairseq_optimizer"* ]]; then
       echo "Running $fastseq_py_test_file with EL Attention"
-      USE_EL_ATTN=1 python $fastseq_py_test_file
+      PYTHONIOENCODING=utf8 USE_EL_ATTN=1 python3 $fastseq_py_test_file
```
Comment (feihugis) [inline tests/run_fastseq_tests.sh:9]: just curious what's `PYTHONIOENCODING` for?

File: fastseq/optimizer/transformers/beam_search_optimizer.py [hunk 15]
```diff
@@ -3,67 +3,138 @@
...
-                                cur_len: int,
-                                pad_token_id: int) -> None:
-    """Copied from fairseq for no_repeat_ngram in beam_search"""
-
-    if cur_len + 1 < no_repeat_ngram_size:
-        # return no banned tokens if we haven't generated no_repeat_ngram_size
-        # tokens yet
-        return [[] for _ in range(num_hypos)]
+@replace(_get_ngrams)
+def _get_ngrams_v2(ngram_size: int, prev_input_ids: torch.Tensor, num_hypos: int, pad_token_id: int = None):
```
Comment (feihugis) [inline fastseq/optimizer/transformers/beam_search_optimizer.py:67]: Is there any annotation for the return type?

File: fastseq/optimizer/transformers/beam_search_optimizer.py [hunk 16]
```diff
@@ -115,915 +185,1009 @@ def _update_beam_size(self, num_beams):
...
-
-            model_specific_kwargs: (`optional`) dict
-                Additional model specific kwargs will be forwarded to the
-                `forward` function of the model.
+        # determine generation mode
+        is_greedy_gen_mode = (num_beams == 1) and (num_beam_groups == 1) and do_sample is False
+        is_sample_gen_mode = (num_beams == 1) and (num_beam_groups == 1) and do_sample is True
+        is_beam_gen_mode = (num_beams > 1) and (num_beam_groups == 1) and do_sample is False
+        is_beam_sample_gen_mode = (num_beams > 1) and (num_beam_groups == 1) and do_sample is True
+        is_group_beam_gen_mode = (num_beams > 1) and (num_beam_groups > 1)
```
Comment (feihugis) [inline fastseq/optimizer/transformers/beam_search_optimizer.py:533]: Are our changes compatible with all these generation modes?

File: tests/run_fastseq_tests.sh (hunk 14 above)
Comment (JulianneKnott) [reply tests/run_fastseq_tests.sh:9]: with the transformers docker, it defaults to ascii which causes the tests to fail on non ascii input

File: fastseq_cli/transformers_generate.py [hunk 17]
```diff
@@ -1,5 +1,7 @@
 """From Huggingface Transformers."""
 
+import sys
+import logging
```
Comment (feihugis) [inline fastseq_cli/transformers_generate.py:4]: use the logging wrapped in fastseq as [here](https://github.com/microsoft/fastseq/blob/main/fastseq/optimizer/fairseq/beam_search_optimizer.py#L24)?

File: fastseq_cli/transformers_generate.py [hunk 18]
```diff
@@ -160,6 +162,9 @@ def generate_summaries_or_translations(
         import fastseq  #pylint: disable=import-outside-toplevel
     fout = Path(out_file).open("w", encoding="utf-8")
     model_name = str(model_name)
+    if model_name == 'prophetnet':
+        model = AutoModelForSeq2SeqLM.from_pretrained(model_name, from_tf=True).to(device)
```
Comment (feihugis) [inline fastseq_cli/transformers_generate.py:166]: Line 173 also uses `AutoModelForSeq2SeqLM`. Could we combine these two?

File: fastseq/optimizer/transformers/beam_search_optimizer.py (hunk 16 above)
Comment (JulianneKnott) [reply fastseq/optimizer/transformers/beam_search_optimizer.py:533]: The beam search updates are only applied for is_beam_gen_mode. The model-specific updates (ie attention) are applied in all cases.

File: examples/gpt2/README.md (hunk 11 above)
Comment (JulianneKnott) [reply examples/gpt2/README.md:11]: Yes. Running each 3x results in: 3.9 3.9 3.8 for bsz=64 and 3.8 3.8 3.9 for bsz=128

File: azure-pipelines.yml [hunk 19]
```diff
@@ -15,21 +15,21 @@ jobs:
       demands:
         - agent.name -equals gpu3
     container:
-      image: adsbrainwestus2.azurecr.io/fastseq:dev-py3 
-      endpoint: fastseq-acr
+      image: huggingface/transformers-pytorch-gpu:latest
```
[the remark this answers is not in the document]
Comment (JulianneKnott) [reply azure-pipelines.yml:18]: yes. it allows both the transformers and fairseq tests to pass in the pipeline

File: fastseq/optimizer/transformers/beam_search_optimizer.py (hunk 16 above)
Comment (feihugis) [reply fastseq/optimizer/transformers/beam_search_optimizer.py:533]: Does it also work with `is_group_beam_gen_mode`?

File: fastseq/optimizer/transformers/beam_search_optimizer.py (hunk 15 above)
Comment (JulianneKnott) [reply fastseq/optimizer/transformers/beam_search_optimizer.py]: added one

File: fastseq/optimizer/transformers/beam_search_optimizer.py (hunk 16 above)
Comment (JulianneKnott) [reply fastseq/optimizer/transformers/beam_search_optimizer.py:535]: no is_group_beam_gen_mode does not use the updates. Should that be added?

File: fastseq_cli/transformers_generate.py (hunk 18 above)
Comment (JulianneKnott) [reply fastseq_cli/transformers_generate.py]: combined

File: fastseq/optimizer/transformers/beam_search_optimizer.py (hunk 16 above)
Comment (feihugis) [reply fastseq/optimizer/transformers/beam_search_optimizer.py:535]: If the optimizations can also work for `group_beam_gen_mode`, we can add it. We can support it in the another PR later if it will  take some time to support it.

Comment (JulianneKnott) [reply fastseq/optimizer/transformers/beam_search_optimizer.py:535]: I think it will take some time to make it work for both, so I think it's best if group_beam_gen_mode goes in another pr

Comment (feihugis) [reply fastseq/optimizer/transformers/beam_search_optimizer.py:535]: Sounds good.

**Review by feihugis — APPROVED** [review]
Thanks @JulianneKnott for the revisions! The PR looks good to me!

File: examples/prophetnet/README.md [hunk 20]
```diff
@@ -53,10 +53,10 @@ bash generate_binary_data_for_prophetnet.sh INPUT_DATA_DIR
 
 - CNN daily mail validation data, NVIDIA-V100-16GB
 
-  |      BatchSize      |       128      |
-  |:-------------------:|:--------------:|
-  | transformers-4.12.0 | 3.4 samples/s  |
-  |  above + fastseq    | 6.3 samples/s  |
+  |      BatchSize      |       128      |       128      |       128      |
```
Comment (feihugis) [inline examples/prophetnet/README.md:56]: Typos in batch size.
